### PR TITLE
Memorify DeflateManagedStream

### PR DESCRIPTION
**Pull request by NewellClark**
This is one of the violations found by [this prototype analyzer](https://github.com/dotnet/roslyn-analyzers/pull/4726). 

Unfortunately, one path of execution calls into [IFileFormatReader](https://github.com/dotnet/runtime/blob/master/src/libraries/System.IO.Compression/src/System/IO/Compression/DeflateManaged/FileFormats.cs), which does not support Span and Memory. I did the same thing that I did with my [CryptoStream PR](https://github.com/dotnet/runtime/pull/47207) and check if the memory is an array, and simply do the copy if it is. 

Unlike with `CryptoStream`, however, `IFileFormatReader` is an internal interface, so we may be able to change it down the road. Might be something worth considering. I can create an issue for that if it's deemed worthwhile.

### CODE REVIEW

**Comment by NewellClark** [comment]
Area label should be `System.IO.Compression`

File: src/libraries/System.IO.Compression/src/System/IO/Compression/DeflateManaged/DeflateManagedStream.cs [hunk 1]
```diff
@@ -234,17 +222,13 @@ private async Task<int> ReadAsyncCore(Task<int> readTask, byte[] buffer, int off
...
 
                     if (bytesRead == 0 && !_inflater.Finished())
                     {
                         // We could have read in head information and didn't get any data.
                         // Read from the base stream again.
-                        readTask = _stream!.ReadAsync(_buffer, 0, _buffer.Length, cancellationToken);
-                        if (readTask == null)
-                        {
-                            throw new InvalidOperationException(SR.NotSupported_UnreadableStream);
-                        }
```
Comment (stephentoub) [inline src/libraries/System.IO.Compression/src/System/IO/Compression/DeflateManaged/DeflateManagedStream.cs:247]: Why did you delete this null check?

Comment (stephentoub) [reply src/libraries/System.IO.Compression/src/System/IO/Compression/DeflateManaged/DeflateManagedStream.cs:247]: Oh, nevermind, you changed to use the overload that's returning a ValueTask.

File: src/libraries/System.IO.Compression/src/System/IO/Compression/DeflateManaged/InflaterManaged.cs [hunk 2]
```diff
@@ -140,16 +142,27 @@ public int Inflate(byte[] bytes, int offset, int length)
                     if (_hasFormatReader)
                     {
                         Debug.Assert(_formatReader != null);
-                        _formatReader.UpdateWithBytesRead(bytes, offset, copied);
+                        // Skip the copy if bytes is actually an array.
```
Comment (stephentoub) [inline src/libraries/System.IO.Compression/src/System/IO/Compression/DeflateManaged/InflaterManaged.cs:145]: This is all dead code.  _formatReader is always going to be null.

Comment (stephentoub) [reply src/libraries/System.IO.Compression/src/System/IO/Compression/DeflateManaged/InflaterManaged.cs:145]: https://github.com/dotnet/runtime/pull/47408

Comment (NewellClark) [reply src/libraries/System.IO.Compression/src/System/IO/Compression/DeflateManaged/InflaterManaged.cs:145]: That's awesome! I think what I'm going to do is wait until #47408 is merged, and then I'll redo this pull-request. That way we'll avoid merge conflicts.

**Comment by NewellClark** [comment]
I'm closing this for now. I'll redo it once [47408](https://github.com/dotnet/runtime/pull/47408) has merged to avoid merge conflicts, and because my implementation broke CI.
